File: exts/jphp-parser-ext/src/main/java/org/develnext/jphp/parser/classes/SourceToken.java
package org.develnext.jphp.parser.classes;

import org.develnext.jphp.core.tokenizer.token.CommentToken;
import org.develnext.jphp.core.tokenizer.token.Token;
import org.develnext.jphp.core.tokenizer.token.expr.ExprToken;
import org.develnext.jphp.core.tokenizer.token.expr.OperatorExprToken;
import org.develnext.jphp.core.tokenizer.token.stmt.StmtToken;
import org.develnext.jphp.parser.ParserExtension;
import php.runtime.annotation.Reflection.Abstract;
import php.runtime.annotation.Reflection.Getter;
import php.runtime.annotation.Reflection.Namespace;
import php.runtime.annotation.Reflection.Signature;
import php.runtime.env.Environment;
import php.runtime.lang.BaseObject;
import php.runtime.memory.ArrayMemory;
import php.runtime.reflection.ClassEntity;

@Abstract
@Namespace(ParserExtension.NS)
public class SourceToken extends BaseObject {
    protected Token token;

    public SourceToken(Environment env, Token token) {
        super(env);
        this.token = token;
    }

    public SourceToken(Environment env, ClassEntity clazz) {
        super(env, clazz);
    }

    @Signature
    public boolean isNamedToken() {
        return token.isNamedToken();
    }

    @Signature
    public boolean isStatementToken() {
        return token instanceof StmtToken;
    }

    @Signature
    public boolean isExpressionToken() {
        return token instanceof ExprToken;
    }

    @Signature
    public boolean isOperatorToken() {
        return token instanceof OperatorExprToken;
    }

    @Signature
    public ArrayMemory getMeta() {
        ArrayMemory r = new ArrayMemory();

        if (token instanceof CommentToken) {
            r.refOfIndex("comment").assign(((CommentToken) token).getComment());
            r.refOfIndex("kind").assign(((CommentToken) token).getKind().name());
        }

        return r.toConstant();
    }

    @Getter
    public String getType() {
        String name = token.getClass().getSimpleName();

        if (name.endsWith("Token")) {
            name = name.substring(0, name.length() - 5);
        }

        return name;
    }

    @Getter
    public String getWord() {
        return token.getWord();
    }

    @Getter
    public int getLine() {
        return token.getMeta().getStartLine();
    }

    @Getter
    public int getPosition() {
        return token.getMeta().getStartPosition();
    }

    @Getter
    public int getEndLine() {
        return token.getMeta().getEndLine();
    }

    @Getter
    public int getEndPosition() {
        return token.getMeta().getEndPosition();
    }
}

File: exts/jphp-parser-ext/src/main/resources/JPHP-INF/sdk/phpx/parser/SourceToken.php
<?php
namespace phpx\parser;

/**
 * Class SourceToken
 * @package phpx\parser
 */
abstract class SourceToken
{
    /**
     * ClassStmt, FunctionStmt, Name, ...
     * @var string
     */
    public $type;

    /**
     * @var string
     */
    public $word;

    /**
     * @var int
     */
    public $line;

    /**
     * @var int
     */
    public $position;

    /**
     * @var int
     */
    public $endLine;

    /**
     * @var int
     */
    public $endPosition;

    /**
     * @return bool
     */
    public function isNamedToken()
    {
    }

    /**
     * @return bool
     */
    public function isStatementToken()
    {
    }

    /**
     * @return bool
     */
    public function isExpressionToken()
    {
    }

    /**
     * @return bool
     */
    public function isOperatorToken()
    {
    }

    /**
     * @return array
     */
    public function getMeta()
    {
    }
}